common/hashing.h

```diff
@@ -0,0 +1,721 @@
+  return hash;
+}
+
+inline auto CarbonHash(Hasher hash, llvm::StringRef value) -> Hasher {
+  return CarbonHash(
+      std::move(hash),
+      llvm::ArrayRef<std::byte>(
+          reinterpret_cast<const std::byte*>(value.data()), value.size()));
+}
+
+inline auto CarbonHash(Hasher hash, std::string_view value) -> Hasher {
+  return CarbonHash(
+      std::move(hash),
+      llvm::ArrayRef<std::byte>(
+          reinterpret_cast<const std::byte*>(value.data()), value.size()));
+}
+
+inline auto CarbonHash(Hasher hash, const std::string& value) -> Hasher {
+  return CarbonHash(
+      std::move(hash),
+      llvm::ArrayRef<std::byte>(
+          reinterpret_cast<const std::byte*>(value.data()), value.size()));
+}
+
+// C++ guarantees this is true for the unsigned variants, but we require it for
+// signed variants and pointers.
+static_assert(std::has_unique_object_representations_v<int8_t>);
+static_assert(std::has_unique_object_representations_v<int16_t>);
+static_assert(std::has_unique_object_representations_v<int32_t>);
+static_assert(std::has_unique_object_representations_v<int64_t>);
+static_assert(std::has_unique_object_representations_v<void*>);
+
+// C++ uses `std::nullptr_t` but doesn't make it have a unique object
+// representation infuriatingly. Turn it back into a real pointer to fix this.
+// This is quite tricky to do in C++. For example, one might think to *just* use
+// the type function below, but that won't work because we'll still at some
+// point need to convert to a `const void*` value in generic code, and there is
+// a *lot* of generic code below this dispatch level. One might then think,
+// fine, *just* use the function that actually maps the value, but that's
+// difficult because it needs to pass along a `const &` in one specialization,
+// but *can't* pass it along on the other because then it would return a `const
+// &` to a local: converting from a `nullptr` to a `const void*` creates a
+// temporary pointer.
+template <typename T>
+struct MapNullPtrTToVoidPtrImpl {
+  using Type = T;
+};
+template <>
+struct MapNullPtrTToVoidPtrImpl<std::nullptr_t> {
+  using Type = const void*;
+};
+template <typename T>
+using MapNullPtrTToVoidPtr = typename MapNullPtrTToVoidPtrImpl<T>::Type;
+
+template <typename T>
+inline auto MapNullPtrToVoidPtr(const T& value) -> const T& {
+  // This overload should never be selected for `std::nullptr_t`, so
+  // static_assert to get some better compiler error messages.
+  static_assert(!std::is_same_v<T, std::nullptr_t>);
+  return value;
+}
+
+// Overload to map `nullptr` values to actual pointer values that are null. This
+// includes returning by-value as the new pointer value will be a temporary.
+inline auto MapNullPtrToVoidPtr(std::nullptr_t /*value*/) -> const void* {
+  return nullptr;
+}
+
+template <typename T>
+constexpr bool NullPtrOrHasUniqueObjectRepresentations =
+    std::has_unique_object_representations_v<MapNullPtrTToVoidPtr<T>>;
```

Comment:
This appears to be the only place using `MapNullPtrTToVoidPtr<T>` -- perhaps `MapNullPtrTToVoidPtr` should be written as an implementation detail of this?